Tietokilta/ilmomasiina
# Port server to TypeScript

The entire application will be ported to TypeScript. This PR will start that with the backend.

The API stays pretty close to previous versions, but it is now _much_ more restricted to what the frontend actually needs. I tested the existing frontend and it seems to work fine with the TypeScript API.

The DB schema is close to previous but will require a migration for existing installations - a script for that will be created at some point. That will be delayed until further code upgrades have been performed later in Tietokilta's website project (including some new features).

Server libraries have been updated to the latest major versions, and as a consequence this update requires Node 12 (current LTS) or 14.

The server has seen some major refactorings that make the port much easier. Mainly, the FeathersJS services have been switched away from the hook model, as it's basically impossible to maintain in TypeScript.

## Testing

To test the system locally:

1. Install a MySQL server. Create a database and a user that has access to it (or just use `root`).
2. Copy `ENV.MD` to `.env` and update MySQL credentials. Everything else should work out of the box.
3. Run:
    ```shell
    git checkout feature/typescript-2
    npm install --also=dev
    npm run compile:client
    npm run start:dev
    ```

## Current status

- Server dependencies upgraded to latest
- All server files ported to TypeScript
- Build working, including autorestart
- Utility scripts deleted (see message for 80c8e7a)
- Builds without linter errors (only warnings for `console`)
- Many bugs fixed in the process
